core/tls-websocket-server/src/lib.rs

```diff
@@ -36,87 +36,127 @@ pub mod sgx_reexport_prelude {
-	InvalidWsAddress(AddrParseError),
-	#[error("TCP bind error: {0}")]
-	TcpBindError(IoError),
-	#[error("Web-socket hand shake error")]
-	HandShakeError,
-	#[error("Web-socket connection already closed error")]
-	ConnectionClosed,
-	#[error("Web-socket connection has not yet been established")]
-	ConnectionNotYetEstablished,
-	#[error("Web-socket write error: {0}")]
-	SocketWriteError(String),
-	#[error("Web-socket handler error: {0}")]
-	HandlerError(Box<dyn std::error::Error + Sync + Send + 'static>),
-}
+pub mod config_provider;
+mod connection;
+pub mod connection_id_generator;
+pub mod error;
+mod tls_common;
+pub mod ws_server;
 
-pub type WebSocketResult<T> = Result<T, WebSocketError>;
+#[cfg(test)]
+mod test;
 
-/// abstraction of a web socket connection
-pub trait WebSocketConnection: Send + Sync {
-	fn process_request<F>(&mut self, initial_call: F) -> WebSocketResult<String>
-	where
-		F: Fn(&str) -> String;
+/// Connection token alias.
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
+pub struct ConnectionToken(pub usize);
 
-	fn send_update(&mut self, message: &str) -> WebSocketResult<()>;
+impl From<ConnectionToken> for Token {
+	fn from(c: ConnectionToken) -> Self {
+		Token(c.0)
+	}
+}
 
-	fn close(&mut self);
+impl From<Token> for ConnectionToken {
+	fn from(t: Token) -> Self {
+		ConnectionToken(t.0)
+	}
 }
 
-/// Handles a web-socket connection
-pub trait WebSocketHandler {
-	type Connection: WebSocketConnection;
+/// Handles a web-socket connection message.
+pub trait WebSocketMessageHandler: Send + Sync {
+	fn handle_message(
+		&self,
+		connection_token: ConnectionToken,
+		message: String,
+	) -> WebSocketResult<Option<String>>;
+}
 
-	fn handle(&self, connection: Self::Connection) -> WebSocketResult<()>;
+/// Allows to send response messages to a specific connection.
+pub trait WebSocketResponder: Send + Sync {
+	fn send_message(
+		&self,
+		connection_token: ConnectionToken,
+		message: String,
+	) -> WebSocketResult<()>;
 }
 
-/// Run a web-socket server with a given handler
+/// Run a web-socket server with a given handler.
 pub trait WebSocketServer {
 	type Connection;
 
-	fn run<Handler>(&self, handler: Arc<Handler>) -> WebSocketResult<()>
-	where
-		Handler: WebSocketHandler<Connection = Self::Connection>;
+	fn run(&self) -> WebSocketResult<()>;
+
+	fn shut_down(&self) -> WebSocketResult<()>;
 }
 
-pub fn run_ws_server<Handler>(addr_plain: &str, handler: Arc<Handler>)
-where
-	Handler: WebSocketHandler<Connection = TungsteniteWsConnection>,
-{
-	let cert = "end.fullchain".to_string();
-	let key = "end.rsa".to_string();
+/// Abstraction of a web socket connection using mio.
+pub(crate) trait WebSocketConnection: Send + Sync {
+	/// Socket type, typically a TCP stream.
+	type Socket: Evented;
+
+	/// Get the underlying socket (TCP stream)
+	fn socket(&self) -> &Self::Socket;
+
+	/// What IO events we're currently waiting for,
+	/// based on wants_read/wants_write.
+	fn event_set(&self) -> mio::Ready;
```

Comment:
I don't think this function name is very descriptive. It does not really help me to understand what this function actually does.

Reply:
It's taken from the mio example - but I agree, I'll try to come up with a better name